pkg/kubelet/status/generate.go

```diff
@@ -29,8 +29,102 @@ const (
 	PodCompleted             = "PodCompleted"
 	ContainersNotReady       = "ContainersNotReady"
 	ContainersNotInitialized = "ContainersNotInitialized"
+	ReadinessGatesNotReady   = "ReadinessGatesNotReady"
 )
 
+// GeneratePodReadyCondition returns ready condition if all containers in a pod are ready, else it
+// returns an unready condition.
+func GeneratePodReadyConditions(spec *v1.PodSpec, conditions []v1.PodCondition, containerStatuses []v1.ContainerStatus, podPhase v1.PodPhase) (podReady v1.PodCondition, containersReady v1.PodCondition) {
+	setReadyConditionsFunc := func(condition v1.PodCondition) {
+		podReady = v1.PodCondition{
+			Type:    v1.PodReady,
+			Status:  condition.Status,
+			Reason:  condition.Reason,
+			Message: condition.Message,
+		}
+
+		containersReady = v1.PodCondition{
+			Type:    v1.ContainersReady,
+			Status:  condition.Status,
+			Reason:  condition.Reason,
+			Message: condition.Message,
+		}
+	}
+
+	// Find if all containers are ready or not.
+	if containerStatuses == nil {
+		setReadyConditionsFunc(v1.PodCondition{
+			Status: v1.ConditionFalse,
+			Reason: UnknownContainerStatuses,
+		})
+		return
+	}
+	unknownContainers := []string{}
+	unreadyContainers := []string{}
+	for _, container := range spec.Containers {
+		if containerStatus, ok := podutil.GetContainerStatus(containerStatuses, container.Name); ok {
+			if !containerStatus.Ready {
+				unreadyContainers = append(unreadyContainers, container.Name)
+			}
+		} else {
+			unknownContainers = append(unknownContainers, container.Name)
+		}
+	}
+
+	// If all containers are known and succeeded, just return PodCompleted.
+	if podPhase == v1.PodSucceeded && len(unknownContainers) == 0 {
+		setReadyConditionsFunc(v1.PodCondition{
+			Status: v1.ConditionFalse,
+			Reason: PodCompleted,
+		})
+		return
+	}
+
+	unreadyMessages := []string{}
+	if len(unknownContainers) > 0 {
+		unreadyMessages = append(unreadyMessages, fmt.Sprintf("containers with unknown status: %s", unknownContainers))
+	}
+	if len(unreadyContainers) > 0 {
+		unreadyMessages = append(unreadyMessages, fmt.Sprintf("containers with unready status: %s", unreadyContainers))
+	}
+	unreadyMessage := strings.Join(unreadyMessages, ", ")
+	if unreadyMessage != "" {
+		setReadyConditionsFunc(v1.PodCondition{
+			Status:  v1.ConditionFalse,
+			Reason:  ContainersNotReady,
+			Message: unreadyMessage,
+		})
+		return
+	}
+
```

Comment:
This is a LOOONG function - could use a lot more comments or breaking up to smaller functions?

Reply:
added comments step by step